### PR TITLE
Refactor/multi threshold confusion matrix

### PR DESCRIPTION
**Pull request by ryandaryl**
To implement a constant-memory version of the AUROC metric, I did the following:
- Modified the ConfusionMatrix metric so that you can give it multiple thresholds as a list or Tensor. In the returned confusion matrix tensor, I've added an extra `threshold` dimension, to accommodate multiple confusion matrices.
- Changed the ROC functions in functional/classification/roc.py, to allow them to take precalculated false-positive and true-positive counts.
- Made ConfusionMatrix the parent class of the AUROC metric.
- Added `reduce_memory` option to AUROC. If `reduce_memory` is True, the following procedure occurs:
    - First, collect a sample of preds / targets, up to `num_batches_to_sample`.
    - Use the sample to decide which thresholds to use for the ROC.
    - For each batch after `num_batches_to_sample`, call ConfusionMatrix update, passing in the preds/targets from the batch, as well as the thresholds.
- Updated AUROC.compute so that, if `reduce_memory` is set to True, it calls ConfusionMatrix compute, and then passes the false-positives and true-positives to the `roc` function to calculate AUROC.

Caveats:

- For backwards compatibility, in ConfusionMatrix, when the caller supplies only a single, float threshold, I've squeezed the `thresholds` dimension. This means the number of dimensions of the ConfusionMatrix output can change depending on the input.
- The original code determines which thresholds to use based on the whole set of preds/targets. To make my constant-memory implementation decide on a set of thresholds, I've had to make it take a sample of the first few batches up to `num_batches_to_sample`. This means the behaviour of my AUROC implementation changes dependent on its state, which could potentially cause problems for people using AUROC.
- The ROC calculations in functional/classification/roc.py assume that the whole set of preds/targets will be passed to the `roc` function. I made `preds` and `target` optional, and I added an option for passing in precalculated false-positives and true-positives. But the resulting code is quite messy. Another option would be to refactor `roc`, and move the parts that don't require `preds` and `target` into their own function.

### CODE REVIEW

**Comment by ryanmseer** [comment]
This is a test comment.

File: torchmetrics/classification/auroc.py [hunk 1]
```diff
@@ -149,13 +160,25 @@ def __init__(
...
-            "Metric `AUROC` will save all targets and predictions in buffer."
-            " For large datasets this may lead to large memory footprint."
-        )
+        if not self.reduce_memory:
+            rank_zero_warn(
+                "Metric `AUROC` will save all targets and predictions in buffer."
+                " For large datasets this may lead to large memory footprint."
+            )
+
+    def _furthest_values(self, input_values, max_count):
```
Comment (ronrest) [inline torchmetrics/classification/auroc.py:173]: I think a docstring explaining what this function does would be helpful. It is not clear to me what it is doing.

File: torchmetrics/classification/auroc.py [hunk 2]
```diff
@@ -118,20 +119,30 @@ def __init__(
         max_fpr: Optional[float] = None,
         compute_on_step: bool = True,
         dist_sync_on_step: bool = False,
+        reduce_memory: bool = False,
         process_group: Optional[Any] = None,
         dist_sync_fn: Callable = None,
+        max_threshold_count = 100,
+        num_batches_to_sample = 4,
```
Comment (ronrest) [inline torchmetrics/classification/auroc.py:126]: These new arguments should be added to the docstring, to explain what they do. Of particular interest might be how `reduce_memory` reduces memory.

**Review by ronrest — CHANGES_REQUESTED** [review]
I can only make a few suggestions regarding the inclusion of docstrings. 

I dont have a good enough grasp on the Metrics to be able to comment on the logic. I feel that Matias should be added as a reviewer. He will have a much better understanding than me about Metrics.

File: torchmetrics/classification/auroc.py [hunk 3]
```diff
@@ -118,20 +119,30 @@ def __init__(
         max_fpr: Optional[float] = None,
         compute_on_step: bool = True,
         dist_sync_on_step: bool = False,
+        reduce_memory: bool = False,
```
Comment (timesler) [inline torchmetrics/classification/auroc.py:122]: This is strictly always a reduction. For a very small dataset and a moderately large threshold count, it will actually increase memory usage. Maybe `fixed_memory` would be better?

File: torchmetrics/functional/classification/confusion_matrix.py [hunk 4]
```diff
@@ -33,23 +33,30 @@ def _confusion_matrix_update(
...
+            minlength = num_classes ** 2
+        bins = torch.bincount(unique_mapping, minlength=minlength)
+        if multilabel:
+            confmat_list.append(bins.reshape(num_classes, 2, 2))
+        else:
+            confmat_list.append(bins.reshape(num_classes, num_classes))
+    confmat = torch.stack(confmat_list)
+    # If a single threshold is provided, return the confusion matrix without the thresholds dimension
+    if isinstance(input_threshold, float):
+        confmat = confmat.squeeze(dim=0)
```
Comment (timesler) [inline torchmetrics/functional/classification/confusion_matrix.py:59]: I think any reasonably efficient implementation of these metrics needs for this code to be vectorised (presumably like we have done it in [seer-algo](https://github.com/seermedical/seer-algo/blob/dcf53db78117401835e7e856ed4ae54b62f5c6db/seeralgo/metrics/metrics.py#L136). The original approach used doesn't lend itself well to doing that since it relies on first modifying the `preds` before using `argmax` (and to vectorise that we'd need to create a huge number of copies of `preds`).

I think we might be best off just implementing a new `ConfusionMatrixCurve` or similar and implementing in the optimal way. Then we can find other places in the existing binned metrics that can take advantage of it

**Review by timesler — CHANGES_REQUESTED** [review]
I think life will be much easier if we start by implementing a new `ConfusionMatrixCurve` (or some better name if you have one) that does what we need efficiently. We'll probably want to make sure it supports multilabel classification as well as both probabilities and logits. Once we have that, it should be able to form the basis for any fixed memory metrics we want, as well as being used by any existing fixed memory metrics

File: torchmetrics/functional/classification/confusion_matrix.py (hunk 4 above)
Comment (ryanmseer) [reply torchmetrics/functional/classification/confusion_matrix.py:59]: Yep I agree it's probably better to leave the original confusion matrix function alone, and make a new one that can handle multiple thresholds.
